drivers/usb/usbport/device.c

```diff
@@ -885,6 +885,76 @@ USBPORT_AbortTransfers(IN PDEVICE_OBJECT FdoDevice,
     }
 }
 
+PUSB2_TT_EXTENSION
+NTAPI 
+USBPORT_GetTt(IN PDEVICE_OBJECT FdoDevice,
+              IN PUSBPORT_DEVICE_HANDLE HubDeviceHandle,
+              OUT PUSHORT OutPort,
+              OUT PUSBPORT_DEVICE_HANDLE * OutHubDeviceHandle)
+{
+    PUSBPORT_DEVICE_HANDLE DeviceHandle = HubDeviceHandle;
+    ULONG TtCount;
+    PLIST_ENTRY Entry;
+    PUSB2_TT_EXTENSION TtExtension = NULL;
+
+    DPRINT("USBPORT_GetTt: HubDeviceHandle - %p\n", HubDeviceHandle);
+
+    *OutHubDeviceHandle = NULL;
+
+    while (DeviceHandle->DeviceSpeed != UsbHighSpeed)
+    {
+        DPRINT("USBPORT_GetTt: DeviceHandle - %p, DeviceHandle->PortNumber - %X\n",
+               DeviceHandle,
+               DeviceHandle->PortNumber);
+
+        *OutPort = DeviceHandle->PortNumber;
+
+        DeviceHandle = DeviceHandle->HubDeviceHandle;
+
+        if (!DeviceHandle)
+            return NULL;
```

Comment:
The `DeviceHandle` passed to this function is always a valid pointer?

Reply:
Yes, we do a USBPORT_ValidateDeviceHandle() check in USBPORT_CreateDevice() before calling USBPORT_GetTt().
But in this line we check the value is not "passed to this function", but the value of the parent DeviceHandle: ```DeviceHandle = DeviceHandle->HubDeviceHandle;```
USBPORT_GetTt() is executed when a USB1.1 device is connected to the USB2.0 controller:
```
     if (Packet-> MiniPortFlags & USB_MINIPORT_FLAGS_USB2 &&
         ! (PortStatus & USB_PORT_STATUS_HIGH_SPEED))
```
It can be any device, including USB1.1 Hub. If this is a USB1.1 Hub, then we need to find the parent USB Hub and this must necessarily be a USB2.0 Hub.
Therefore, ```if (! DeviceHandle)``` is an error.